Added selectable "ordinary" or "persistent" behavior of variables to "Dummy Device"

BOBAH1248 opened the pull request:
Like some of the other guys, I ran into the need to restore a device to a previous state after a power failure.
I tried several solutions, but I ran into their complexity, or the requirement for an extension of the rules language, and almost always that they did not fit into the standard build of the project.
Finally, I managed to come up with a very simple solution that does not take up much memory, computing power, fits perfectly into the settings of existing devices and even works great offline.

All a guy who uses "Dummy Device" to store GPIO states (as it was described in guidelines), in order for his device to restore to the state before power loss, needs to do is on the device settings page to click the "Restore state after power lost" checkbox

svhelper committed:
Added selectable "ordinary" or "persistent" behavior of variables to "Dummy Device"

TD-er commented on src/_P033_Dummy.ino:
```diff
@@ -62,8 +62,72 @@ boolean Plugin_033(uint8_t function, struct EventStruct *event, String& string)
+        }
+        if (success) {
+          //SaveTaskSettings(event->TaskIndex);
+          String res = SaveSettings();
```
This is going to wear out the flash extremely fast when the checkbox is checked by someone who doesn't know what it may entail.

Also the check for a floating point value is done wrong here, as there are floating point values which will never match, like `NaN`.
When comparing a NaN with itself, its result is always false. That's how you can detect a value is NaN.
So within 100 sec, you will run into the error message stating you've exceeded the daily limit of flash writes.

BOBAH1248 replied:
> This is going to wear out the flash extremely fast when the checkbox is checked by someone who doesn't know what it may entail.

Exactly, for this reason I place it to `PLUGIN_ONCE_A_SECOND` case.
But in same use-cases, it will be better to do it in `PLUGIN_WRITE` instead.


> Also the check for a floating point value is done wrong here, as there are floating point values which will never match, like `NaN`.

You right, in this case, I also preferred to use single binary comparison as snapshot (`memcmp` + `memcpy`) instead of in cycle check variable-by-variable according to type.
It will avoid `NaN` and similar issues.

TD-er replied:
> Exactly, for this reason I place it to `PLUGIN_ONCE_A_SECOND` case.

Use case: Store a NaN in the dummy
Result: Every second the settings file will be saved.

Or some rule storing temporary values in a dummy instead of variable. This may also change multiple times a second => every second the settings will be saved.

BOBAH1248 replied:
That's right

But let's look at the whole situation:
1. Dummy Device cannot update its variables on its own
2. Maybe I'm wrong, but its variables can be changed using rules (at least I haven't come across other options)
3. Self-writing rules requires learning the language and architecture features (not only external, but also internal)
4. A person who design rules on his own is forced to debug them and monitor the correct execution of the logic for some time, so he may enough quickly ran in to the described restrictions
5. A person who uses someone else's rules should at least read the description, requirements for it, as well as restrictions
6. Perhaps the most important thing: a person when used rules, he does solely at his own peril and risk

Therefore, since the Dummy Device itself is safe, imposing restrictions on its use imposes restrictions on its functionality as well as system at whole.

But we digress, this code is outdated and has been changed. Could you share your thoughts on the final version?

TD-er replied:
You set values to a dummy using `taskvalueset` or `taskvaluesetandrun`
The "run" is actually a call to `PLUGIN_READ`, which on success will 'flush' its values to the assigned controllers.

BOBAH1248 replied:
OK, if we will not pay attention to the fact that there are only 4 controllers and adding new ones is, to put it mildly, problematic, what happen if controller is not ready to process it?
Or if controller is "storage" and made over flash, it will avoid killing of flash-chip during saving every second by few bytes?

BOBAH1248 replied:
Another point: if "storage controller" will be implemented, it may be integrated to the current build configurations? For example, ESP_Easy_mega_20220511_normal_ESP8266_1M ?

TD-er replied:
There is a storage controller called the "cache controller", which does append to .bin files on the filesystem.
This way it doesn't wear out the flash.
It is meant to log samples while offline and has been used in some test systems for years now.

But that's not the problem here. The problem is that your code tries to save the same settings file over and over again and that should be dealt with, which we can.

BOBAH1248 replied:
Probably you told about C016
I found that it enabled for `PLUGIN_BUILD_MAX_ESP32` (via PLUGIN_SET_MAX)
And available at "Maximized build definition for an ESP(32) with 16MB Flash and 4MB sketch partition".

Is there a way to achieve something similar on more modest chips?
And how soon should this be expected?

BOBAH1248 replied:
Just now I add line
`static_assert(sizeof(Settings) < 4*1024, "Settings large than 4k");`
And try to compile - all done fine even for `mega normal_ESP8266_1M`
Then I check the `SaveSettings()`, that saves `Settings` to the "config.dat" file at `index` = 0 (probably it's position from the begin of file / `0, SeekSet`)
So, at low-level for spi-flash 25qXX series all fine - single sector will be touched.
I just can't see any reason not to use this ugly but usable and helpful code

TD-er replied:
You can see the layout of the files graphically layout on the sysinfo page after giving the command meminfodetail.
(not on "limited" builds like the "testing" builds)
On SPIFFS partial updates in a file can be written relatively fast, but on LittleFS it takes a lot more effort.

The Cache controller is working just fine on ESP8266, as it was initially written for that platform.
But it is not possible to read the data back from within ESPEasy. There is a htm file present to help decode the bin files into .csv for further processing.

BOBAH1248 replied:
Thanks a lot for explanations

BOBAH1248 replied:
I found technical description of SPIFFS internals
https://github.com/pellepl/spiffs/blob/master/docs/TECH_SPEC

It looks like, that both operations of "write" and "append" will allocate new logical blocks.
The LittleFS do the same.
This this approach is called "copy-on-write", and provides "wear leveling" for flash memory.

The main differences between SPIFFS and LittleFS are:
- LittleFS faster due to optimized storage structure for fast searching of indexes
- LittleFS has high robustness - it may safely rollback uncommitted changes (include firmware failure, power lost and so)
- LittleFS detects bad physical sectors and avoid they usage

So, due to SPIFFS not optimally stored metadata (file indexes, allocations and so), on the large filesystem area, after number of writings, its work will very slowly. And more slowly if file will be large
see https://github.com/esp8266/Arduino/issues/4291

From my point of view, so that operations with files on SPIFFS do not slow down over time, file fragmentation should be avoided. In other words, it is desirable to rewrite entire files.
Unfortunately, with frequent operations, this will significantly shorten the life of the flash memory, and not its individual sections, but the whole (remember the built-in "wear leveling"?).

BOBAH1248 replied:
It turns out that the best way to restore the previous values ​​is:
A. either storage in an external chip that wear resistant (like FRAM)
B. either determine the moment of power outage (brownout detection), and store the data in a chip that does not fit the description (A)

That's just unfortunately, ESPs are mostly equipped with 3.3V flash memory, which works from 2.7V, while ESP continues to work at 1.8-2.5V. I.e. at the time of power failure, flash memory is not accessible.
It turns out that option (B) can be used if the ESP is equipped with 1.8V flash memory.

TD-er replied:
> I found technical description of SPIFFS internals https://github.com/pellepl/spiffs/blob/master/docs/TECH_SPEC
> 
> It looks like, that both operations of "write" and "append" will allocate new logical blocks. The LittleFS do the same. This this approach is called "copy-on-write", and provides "wear leveling" for flash memory.
> 
> The main differences between SPIFFS and LittleFS are:
> 
> * LittleFS faster due to optimized storage structure for fast searching of indexes
> * LittleFS has high robustness - it may safely rollback uncommitted changes (include firmware failure, power lost and so)
> * LittleFS detects bad physical sectors and avoid they usage
> 

[...]

You're missing one very important difference between LittleFS and SPIFFS.
LittleFS does not (!!!) support partial updates of a file.
For example the ESPEasy settings file, which is essentially a concatenation of blocks of fixed size per task/controller, is just about the worst possible layout for updating the settings of a single task or something else not at the end of the file.
It means that everything from the point of changes has to be rewritten on LittleFS.

Thus saving something related to a task may easily take several seconds to save on LittleFS.

N.B. I will (soon) change how settings are stored on LittleFS, purely for this reason, but that needs quite a lot of changes to add a transparent layer for storing smaller chunks and combining them into a complete file again either when downloading the file, or after some time (or reboot)

If you need to save things often on either FS, it is best when you use a very small file for this. => Calling `SaveSettings()` is the wrong approach.

And as I said, this should be done using a plugin command (thus via `PLUGIN_WRITE`) and not in a periodical loop.

BOBAH1248 replied:
I have seen user complaints on speed on LittleFS about file updating in the middle of file (include your).
And because ESPeasy based on SPIFFS, I stopped discussing LittleFS further.

Now I was only collecting information about existing solutions, solutions to which you plan to switch.
And I also think about the adequate use of both cases.

In any case, the use of `SaveSettings()` can be tried to be justified only by the size of the code. I didn't like it at first.

BOBAH1248 replied:
To do everything right needed:
1. for compatibility with FS and NAND-flash (SPIFFS, LittleFS or other) need to store the data of each task in a separate file
2. those files should be overwritten in its entirety, which will allow slower FS like SPIFFS to not significantly reduce the speed of work
3. the ideal case for LittleFS would be if the data size does not exceed 104 bytes, in which case only one low-level I\O operation will be required to write/overwrite a file
4. do not scan memory for changes, but use `PLUGIN_WRITE` notification
5. limit the number of writes to a file by time (e.g. daily limit)
6. be able to somehow fit into the available amount of program memory in order to pass automatic compliance tests

Do you have any comments or additional wishes?

TD-er replied:
nr 5. is already implemented if you just use the file handling functions in the ESPEasyStorage source code files.
And appending to a file is more light-weight upto a certain level (upto 8k). This is a very simple way to balance between speed, code complexity and flash wear. (for both SPIFFS and LittleFS)
You can call "seek" to go to the end of a file minus the expected offset for reading the blob you need to read.
Only thing to keep in mind is that on an 1M node, the filesystem size is only 128k, which does not allow for a lot of slack, so better keep it to 1k or 4k max and only store it in 1 file.

There is one thing to check/consider. 
Opening a file on a larger file system (e.g. 8M LittleFS, but large SPIFFS is even worse) does take time.
On 8M LittleFS this can be as much as 20-30msec.
However keeping a file with write bit open, may cause the cache not to be written.
This can be an optimization as it allows for multiple writes (e.g. calls to several tasks to save and store them in the same file) to be written at once, but it should be flushed every now and then to make sure it is actually written.

Another thing I just realized is that you only need this for load at boot and perhaps only for a few tasks. So maybe it doesn't need to be part of the dummy plugin, but it can also be a variant on `TaskValueSet` command, like `TaskValueStore`, which does store it as described and the recovers it automatically at boot (before even handling the System#boot event).
Also on deleting the task, the values stored should be removed.

BOBAH1248 replied:
Thanks for sharing your thoughts

Good point about `TaskValueStore`

Did you read this?
https://github.com/littlefs-project/littlefs/blob/master/DESIGN.md#files

The "inline files" looks very interesting - up to 104 bytes of data file will be stored in metadata (with file description) instead of allocation of new block. Update of such file will just replace the metadata.
When used appending to a file, not only data should be concatenated to already stored data, but updated new size of file in the metadata,. So it requires at least on single write operation more than replacing of "inline file" entirely.

TD-er replied:
> When used appending to a file, not only data should be concatenated to already stored data, but updated new size of file in the metadata,. So it requires at least on single write operation more than replacing of "inline file" entirely.

Well yes and no...
Yes it needs to update the metadata, but at least SPIFFS and I can't imagine LittleFS doesn't, does pull some clever tricks here to only change '1' into '0', thus no need for a block erase. Such an update -without the need for a block erase- can be considered a "non-write" as it is not further aging/wearing the flash.

I will read your link about the metadata stuff and "inline files" to see how it can be used to improve ESPEasy write speed on updating task settings.
My initial thought was to split the settings file into a (hidden) folder with smaller files per task and only when the file needs to be downloaded, it should be reconstructed into a larger file on the fly while streaming it to the user.

BOBAH1248 replied:
The "inline files" is part of LittleFS, not SPIFFS.
Both of File Systems (FS) doesn't erases flash, but updates (i.e. only change '1' into '0'). Exactly except case of formatting and garbage collection / "shrinking". This operation performed very rarely, probably first start should be after 50-80% of memory area will be used.

Regarding of appending data to file, i.e. "writing on bytes over 0xFF" does not age the flash.
But when data appended to the file, the file size will be changed.
How FS can say user the actual size of file?
Right, this information stored in the metadata of FS.
So, if you will increase size of file, it must be displayed in the file descriptor (metadata).
SPIFFS based on great thing (changing "1" to "0" without aging of flash). But changing of a "data size" field in the file descriptor structure may require changing some bits from "0" to "1" (for example, file was 8 bytes length, then came 9 bytes)

For this reason, the SPIFFS has type "Journalling File System" (JFS) - all changes of metadata will be written as new journalling record. Even LittleFS is not JFS, but has similar behavior for storing of metadata.

As we can see, changing of data with or without file size changing, requires updating of file descriptor/metadata (new location of data or actual file size, etc).
So, any writing operations will update 2 pages on flash - file data area and metadata area.

The LittleFS may store file data within metadata in "journalling stream", if data too small (less 104 bytes).
So instead of deletion of old file, releasing of data area, allocating new data area, updating and creating of new journalling record, the LittleFS will write just put new journalling record that contains new data of file. I.e. single I/O operation.

But this trick is available for files with ultra small data size.

Conclusion: appending requires at least 2 writing operations, overwriting of file with small data requires 1 writing operation.

PS: here all 'writing operations' will not erase flash so will not age it.

BOBAH1248 replied:
The "inline files" may be unusable for storing of config files.
May be except case when you going to use "hidden folder" as ".ini"-file (or single-level xml/json), where keys will be filename, and content will be filedata
Or if use sub-subdirs, you may create multi-level xml/json "file" ;)

Sorry for long read

BOBAH1248 replied:
Purely out of curiosity, I decided to see what, where and how many SPIFFS does when writing files.
For these purposes, I made a test log, which in memory only adds 3-6 uint32_t to a large one-dimensional array at the pointer to its tail and increments it. The log gets the time in milliseconds, a pointer to the text description of the log line and up to 3 parameters (casted to uint32_t).
As a result, the log turned out to be consuming a lot of RAM, but extremely fast. For me - this is what needed for real-time tests.

The test itself consists of:
- opening SPIFFS and formatting (results are omited in the logs)
- 3 iterations of creating a file for appending a string of 13 bytes (at #zero try the file does not yet exist)
- 3 iterations of creating a file to overwrite a string of 13 bytes (like above, at #zero try the file does not yet exist)

In the attachment there are logs, where the first column is the time in milliseconds, the rest are the name of the function and its slightly revised parameters, which are more convenient for perception.

I share results if anyone interested in such
[spiffs_test.txt](https://github.com/letscontrolit/ESPEasy/files/8692800/spiffs_test.txt)

TD-er replied:
Make sure not to draw incorrect conclusions about speed regarding writes to filesystems on embedded systems.
It differs a lot when the file system gets fragmented and the size of the filesystem also makes a huge difference in speed.
For SPIFFS I noticed that appending to files also gets slower when the file grows. Over 24k it got significant slower.

BOBAH1248 replied:
I apologize if your remark was sent to me, but if you read it carefully, I did not voice any conclusions absolutely.
If it was intended for those who will use the log in isolation from the test itself, preconditions, conditions and so, then yes, your comment will be useful to not to do something rash.

BOBAH1248 replied:
After analyzing the logs, I'm ready to share my findings regarding SPIFFS:
1. Creating or opening a file requires many read operations to find the file itself
2. Reopening a file whose data is available yet in the cache (RAM) does not require re-reading the flash
3. The operation of appending to the file does not require the [re]allocation of new pages, even to update the metadata (update file size)
4. The operation of overwriting a file requires deleting it, releasing its pages and allocating new pages for data and metadata (requires additional 5-6 writing operations)
5. The operation of overwriting a file consumes free pages (erased) faster than when appending to the end, which is why the time to start the "garbage collection" operation (to erase unused pages) will come faster, which means that flash wear will occur faster
6. Due to frequent update of the same page, SPIFFS can only be used on NOR-flash and not on NAND-flash (where ECC will be broken after second-third writing)

That's all - the most of those conclusion points are already well known by the most of developers.

The only useful thing I saw here is that when appending, the number of write operations is performed not 2, but 6 times.
Which, from my point of view, is bad, it turns out 3 times more likely to lose data due to failures (for example, power lost)
According to the LittleFS technical description, they placed special emphasis on the atomicity of the metadata update operation.

Perhaps it makes sense to check this statement as well.

uzi18 replied:
@BOBAH1248 maybe SPIFFS alocates some pages for appending to file

BOBAH1248 replied:
@uzi18 probably no, lets check logs:
```
       0 --- Test 'append #0' 13 bytes
       0 INFO: file = SPIFFS.open("/test_a.txt", "a")    -- create new file case
...
      16   - spiffs_hal_write: page=2992; offset=0; size=2    -- (A) page updated after every appending
      16   - spiffs_hal_write: page=2993; offset=0; size=46   -- (B) not sure, probably file description (name and index of datapage
      17 INFO: file.write("Hello world!\n")==13   -- nothing to flash, all data in the cache RAM
      17 INFO: file.close()
      17   - spiffs_hal_write: page=2992; offset=2; size=2   -- (A) again updated, but with new offset, looks like it size of file
      17   - spiffs_hal_write: page=2994; offset=0; size=5   -- (C) allocate page for file data
      18   - spiffs_hal_write: page=2994; offset=5; size=13   -- (C) exactly file data, this size as mark =)
      18   - spiffs_hal_write: page=2993; offset=0; size=256  -- (B) again touch full page of "probably file description"

      19 --- Test 'append #1' 13 bytes
      19 INFO: file = SPIFFS.open("/test_a.txt", "a")   -- file in the cache yet, do nothing
      19 INFO: file.write("Hello world!\n")==13    -- again put the cache in RAM
      19 INFO: file.close()
      19   - spiffs_hal_read : page=2994; offset=0; size=256  -- (C) read file data
      19   - spiffs_hal_write: page=2994; offset=18; size=13  -- (C) append additional 13 bytes (now here: 5 bytes metadata + 13 bytes = 18 offset) 
      20   - spiffs_hal_write: page=2995; offset=0; size=256   -- (D) not sure what is it, but the next time will be updated 1 byte only
      21   - spiffs_hal_write: page=2992; offset=4; size=2   -- (A) again updated, but with new offset, looks like it size of file
      21   - spiffs_hal_write: page=2995; offset=4; size=1   -- (D) ???
      21   - spiffs_hal_write: page=2992; offset=0; size=2   -- (A) again updated, but the 1st "index" (invalidate?)
      21   - spiffs_hal_write: page=2993; offset=4; size=1   -- (B) again touch page of "probably file description"

      22 --- Test 'append #2' 13 bytes
      22 INFO: file = SPIFFS.open("/test_a.txt", "a")    -- 3th step of test
      22 INFO: file.write("Hello world!\n")==13   -- again update cache
      22 INFO: file.close()
      23   - spiffs_hal_write: page=2994; offset=31; size=13  -- (C) append additional 13 bytes (now here: 5 bytes metadata + 13 bytes + 13 bytes = 31 offset) 
      23   - spiffs_hal_write: page=2996; offset=0; size=256   -- (E) behavior similar to (D)
      24   - spiffs_hal_write: page=2992; offset=6; size=2    -- (A) again updated, but with new offset, looks like it size of file
      24   - spiffs_hal_write: page=2996; offset=4; size=1    -- (E) ??? (again like D)
      24   - spiffs_hal_write: page=2992; offset=4; size=2    -- (A) again updated, but the 2nd "index" (invalidate?)
      24   - spiffs_hal_write: page=2995; offset=4; size=1    -- (D) ??? (like above)
```

As we can see, every appending operation allocate additional page (2995, 2996), but if look at writing offset and size, new data appended to already allocated during file creation (at "#zero" step)

The usage of (D),(E) (2995, 2996) is questionable, for clarifications the algorithm should be analyzed  (by back trace or static analysis)

BOBAH1248 replied:
> ...
> The usage of (D),(E) (2995, 2996) is questionable, for clarifications the algorithm should be analyzed (by back trace or static analysis)

Looks like (B), (D), (E) (2993, 2995, 2996) pages are the new|updated file descriptor - at previous step it allocate and validated (by writing 1 byte), then at next step it invalidated (again by writing 1 byte at the same position)

TD-er commented:
This is not the way to do it.
Saving all settings is a lot of overkill to save a few bytes.
Also it is better to add a command for it to explicitly save a value, so at least a user needs to think about when to write to flash instead of possibly writing it every second. (or more)

BOBAH1248 commented:
> Also it is better to add a command for it to explicitly save a value, so at least a user needs to think about when to write to flash instead of possibly writing it every second. (or more)

Agreed. it will be better in same use-cases, it will be better to do it in `PLUGIN_WRITE` instead.

> Saving all settings is a lot of overkill to save a few bytes.

It depend on what `storage` used under hood.
For example, if user going to use `FRAM` or `PSRAM` it won't have any effect.

But, for 99% cases it will be NOR-flash (SPIFFS).
And at low level, to update some bytes, nor-flash driver should preform: read-modify-erase-write for minimal erasable chunk - `sector`, that for most of esp devices has 4KB size
So, even we going to update a single byte, we must:
1. read 4KB
2. update this byte at offset
3. erase sector
4. write 4KB back to flash

Sorry, but at low-level content I don't see any issues, while settings fit to 4k page.

Any way, the nor-flash driver may at step 2 check if it made changes in the sector, and when is "not", it should skip steps 3-4

Exactly the same should be done driver of nand-flash.
In case of FRAM, PSRAM or ~~EEDATA~~ EEPROM, the low-level driver may update the bytes directly on address space of "storage chip".

In other words, the "the daily limit" is not related to this patch, but device configuration (for FRAM, there no daily limits, right?)

> This is not the way to do it.

Agreed, this patch is not solution, but "crutch".
But before creating of this pull request, I carefully check contribution polity for this project, state of source code, and came to the conclusion that crutches are the usual state for the project ;) - there a lot of code that for compatibility didn't changed yet and probably will not

svhelper committed:
To avoid `NAN` issue, changed to binary comparison

svhelper committed:
To reduce the program memory usage, removed useless logs

svhelper committed:
Added compile-time check of "storage" compatibility

svhelper committed:
Added a warning to the user about the limitations of this feature on the WebUI
And due to memory limits, made in 2 variants

BOBAH1248 commented:
As you noticed above:
- I added warnings to the user on the WebUI
- replaced the comparison of data from according to types to a regular "snapshot" (with binary comparison "as is")

And again ran to program memory limits for some builds and impossibility to made something more usable than just "workaround for some cases".

BOBAH1248 committed:
Merge branch 'mega' into mega

BOBAH1248 committed:
Merge branch 'letscontrolit:mega' into mega

BOBAH1248 commented:
@TD-er, I apologize for the delay. I have long wanted to close this request, but I did not have free time.

However, I recently discovered that ESP-IDF has some interesting features, functions prefixed with nvs_. The Arduino library "Preferences" is based on them. In general, this is an API for storing key + data pairs in the "hidden" area of spi-flash. The good news - they have wear leveling (because they use journaling mode for storing key pairs).
I wanted to take advantage of these features to solve the problem raised in this request. But as it turned out, this esp8266 package does not have this functionality.
"It's a shame, annoying, but okay" ;)

Then, I wondered, how does SPIFFS store empty files?
I ran several tests, read the source files, and this is what I found: it turns out that empty files have only names, i.e. take up space only in the "directory" area (pages for file content are not allocated). Also, SPIFFS can store some metadata in the same structure (file descriptor). But unfortunately, in the current project, the metadata size is set to zero. That is, if the metadata were allocated, then SPIFFS, to some extent, could be used on a par with LittleFS and its "inline file".
In general, "and again miss"

But nevertheless, it led me to the idea that it is possible to store some value in the file names (for example, such a file name "/mycfgvalue/1234.5678" or "/mycfgvalue=1234.5678"). This approach is somewhat similar to the same "inline file". The only limitation is the filename length of 31 characters.
I made additional tests:
- created empty files SPIFFS.open("/test_c0=0", "w") + file.close()
- renamed files SPIFFS.rename("/test_c0=0", "/test_c0=10")
- deleted and recreated

All tests showed that SPIFFS:
- allocates only one new page when creating an empty file
- releases this single page on deletion
- and when renaming does both operations described above

@TD-er, How do you think, if instead of what I did earlier, take your advice and create 2 "internal commands" EedataWrite and [EedataRead#...], whose first argument is the file name

For example, toggle a relay by a button with saving the state between power failures (the file name here is "dummy_var1"):
```
on System#Boot
  TaskValueSet,dummy,var1,[EedataRead#dummy_var1]
endon

on dummy#var1 do
  gpio,14,[dummy#var1]
  EedataWrite,dummy_var1,[dummy#var1]
endon

on button#State do
  if[dummy#var1] == 1
    TaskValueSet,dummy,var1,0
  else
    TaskValueSet,dummy,var1,1
  endif
endon
```

TD-er commented:
I'm not sure about how SPIFFS handles filenames regarding wear level.
Typically the file names don't change that often, only the content of a file does.
So I can imagine this is less optimized for wear levelling.

One could create a special file, in which the first byte(s) keep track of how many writes there have been to speed up skipping to the right bits.
This counter is then a bitmap where bits are changed from "1" to "0"
And then append to the file (or rewrite a block which was initialized with all 0xFF values at start)

N.B. on the ESP8266 we still have an EEPROM area we don't (yet) use, so this can also be done there using the eeprom Arduino library.
However I plan on maybe using it to store WiFi credentials for a smaller 2-step OTA bin file.

BOBAH1248 commented:
> I'm not sure about how SPIFFS handles filenames regarding wear level.

This information clearly described in the [SPIFFS doc](https://github.com/pellepl/spiffs/blob/master/docs/TECH_SPEC)
Any way, I'll try to summarize what's in there for those who are interested:
1. all physical flash memory segments (minimal items that may be erased) divided on small logical pages
2. SPIFFS allocates logical pages and put to it information about content (directory or data)
3. [directory pages](https://github.com/pellepl/spiffs/blob/master/src/spiffs_nucleus.h#L492-L512) contains not only [file names](https://github.com/pellepl/spiffs/blob/master/src/spiffs_nucleus.h#L506-L507), but [file size](https://github.com/pellepl/spiffs/blob/master/src/spiffs_nucleus.h#L502-L503) ( [it also described in tech_spec](https://github.com/pellepl/spiffs/blob/master/docs/TECH_SPEC#L167) )
4. even during [appending to file](https://github.com/pellepl/spiffs/blob/master/src/spiffs_nucleus.c#L1224) it will [update the directory entry](https://github.com/pellepl/spiffs/blob/master/src/spiffs_nucleus.c#L1455) using [spiffs_object_update_index_hdr](https://github.com/pellepl/spiffs/blob/master/src/spiffs_nucleus.c#L1019-L1033)  and [write to the new page the "unchanged filename" with "file size"](https://github.com/pellepl/spiffs/blob/master/src/spiffs_nucleus.c#L812-L871)


> One could create a special file, in which the first byte(s) keep track of how many writes there have been to speed up skipping to the right bits.
> This counter is then a bitmap where bits are changed from "1" to "0"
> And then append to the file (or rewrite a block which was initialized with all 0xFF values at start)

With raw NOR-FLASH this will work, but not with SPIFFS


> N.B. on the ESP8266 we still have an EEPROM area we don't (yet) use, so this can also be done there using the eeprom Arduino library.
> However I plan on maybe using it to store WiFi credentials for a smaller 2-step OTA bin file.

It's not good idea if you plan to change data frequently - [the EEPROM library has not wear level](https://github.com/esp8266/Arduino/blob/master/libraries/EEPROM/EEPROM.cpp#L134-L135)

TD-er commented:
> It's not good idea if you plan to change data frequently - [the EEPROM library has not wear level](https://github.com/esp8266/Arduino/blob/master/libraries/EEPROM/EEPROM.cpp#L134-L135)

That I know, that's why I mentioned a low-level approach to be easy on the number of erase cycles.

BOBAH1248 commented:
Let's get back to the topic - what do you think about storing small data in filenames, which will allow you to store a key + data pair with minimal overhead?
This solution is more like a hack, so if you are against any hacks in your project, then we should stop any discussion of the ins and outs of this trick.

TD-er commented:
The problem I have with such a 'hack' of storing information in the file name, is that it is very specific to SPIFFS file systems.
On the ESP32 we don't need it as there is NVS storage.
So why not just implement it for ESP32 using NVS and then later we could "backport" the functionality to ESP8266, perhaps by implementing the same behavior but then use some unused space on the flash.

If we make the commands generic enough, then later it can also be used to store it into actual EEPROM (or FRAM) chips.

In other words, I think  it should be implemented using whatever is already present and then we might consider "hacks" to add it to platforms that don't support these features. But for sure let's not start by implementing hacks and then see how it can be mapped to more elegant features.

BOBAH1248 commented:
As I wrote above, the ESP8266 already supported NVS
But with another package - with RTOS https://github.com/espressif/ESP8266_RTOS_SDK/tree/master/components/nvs_flash

TD-er commented:
Yeah, but RTOS on ESP8266 is not going to work as it leaves way too little RAM to work with.

BOBAH1248 commented:
The problem not in RTOS itself, but in number of RTOS objects, like threads that allocates large stack per every task due to unoptimized application, calls and libraries.
In other words, when use single task (as well single thread) the RTOS will not require large memory

BOBAH1248 commented:
Any way, I did't see that Espressif support ESP8266_RTOS_SDK for Arduino compliance

BOBAH1248 commented:
Due to the fact that specially reserved task settings are stored inside the task description (similar to how SPIFFS stores the file size in the file descriptor), they have a huge size and due to lack of functions for partial update of spi flash, the proposed solution looks awkward.
